syntax= "proto3";

service Market{
  rpc registerSeller(registerSellerReq) returns (registerSellerRes){};
  rpc sellItem(sellItemReq) returns (sellItemRes){};
  rpc addProduct(Product) returns (void){};
  rpc updateProduct(Product) returns (void){};
  rpc deleteProduct(Product) returns (void){};
  rpc viewProduct(void) returns (Product){};
  rpc searchProduct(void) returns (Products){};
  rpc addRating(Rate_an_item) returns (Product){};
  rpc displayProducts(Seller) returns (Products){};
  rpc buyProduct(Products) returns (void){};
  rpc addtoWishlist(Product) returns (Products){};
  rpc NotifyClient(void) returns (void){};


}

message void{

}

enum Status {
  SUCCESS = 0;
  FAILURE = 1;
}

enum Category {
  ELECTRONIC=0;
  FASHION=1;
  OTHERS=2;
}

message Buyer{
  string UUID = 1;
  string address = 2;
}
message Seller{
  string UUID = 1;
  string address = 2;
  repeated Product products=3;
}

message Sellers{
  repeated Seller sellers = 1;
}

message Rate_an_item{
  Buyer buyer=1;
  int32 rating=3;
  string review=4;
}

message Product{
  string name=1;
  float price=2;
  int32 quantity=3;
  string description=4;
  string seller_address=5;
  string seller_UUID=6;
  Seller seller = 7;
  Category category =8;

}

message Products{
  repeated Product products = 1;
}

message registerSellerReq{
  string address = 1;
  string uuid = 2;
}

message registerSellerRes{
 Status status = 1;
}


message sellItemReq{
  string name = 1;
  int32 quantity = 2;
  string description = 3;
  string sellerAddress = 4;
  float price = 5;
  string sellerUUID = 6;
  Category Category = 7;
}

message sellItemRes{
  string productUUID = 1;
  Status status = 2;
}